killerbee/zbwardrive/scanning.py

```diff
@@ -1,147 +1,337 @@
+    tries_limit = 5 # How many retries to give a zigbee device to sync
+    scanners = []   # Stored information about each Scanner class we spawn
+    channels = multiprocessing.Queue() # Keeps track of channels
+
+    # Add the channels to the queue
+    for i in range(11,26):
+        channels.put(i)
+
+    # Sync the devices and init the Scanners
+    for device in devices:
+        log_message =  "Creating {}".format(device[0])
         if verbose:
-            print 'Injecting a beacon request on channel %d.' % channel
-        try:
-            beaconinj = beaconp1 + "%c" % seqnum + beaconp2
-            kbscan.inject(beaconinj)
-        except Exception, e:
-            raise Exception('Unable to inject packet (%s).' % e)
-
-        # Process packets for staytime (default 2 seconds) looking for the beacon response frame
-        endtime = time.time() + staytime
-        nonbeacons = 0
-        while (endtime > time.time()):
-            recvpkt = kbscan.pnext() #get a packet (is non-blocking)
-            # Check for empty packet (timeout) and valid FCS
-            if recvpkt != None and recvpkt['validcrc']:
-                #if verbose: print "Received frame."
-                newNetworkChannel = doScan_processResponse(recvpkt, channel, zbdb, kbscan, verbose=verbose, dblog=dblog)
-                if newNetworkChannel != None:
-                    startCapture(zbdb, newNetworkChannel, gps=currentGPS, dblog=dblog)
-                    nonbeacons = 0 # forget about any non-beacons, as we don't care, we saw a beacon!
-                    break          # made up our mind, stop wasting time
-                elif agressive:    # we may care even though it wasn't a beacon
-                    nonbeacons += 1
+            print log_message
+        logging.debug(log_message)
+
+        # Create Scanner
+        kill_event = multiprocessing.Event()
+        channel = multiprocessing.Value('i',0)
+        kbdevice = create_device(
+            device[0], verbose=verbose, timeout=timeout,
+            tries_limit=tries_limit)
+        scanner_proc = Scanner(
+            kbdevice, device[0], channel, channels,  verbose,
+            currentGPS, kill_event, output, 
+            scanning_time, capture_time)
+
+        # Add scanner information to scanners list
+        s = {}
+        s["dev"] = kbdevice
+        s["devstring"] = device[0]
+        s["channel"] = channel
+        s["proc"] = scanner_proc
+        s["kill"] = kill_event
+        scanners.append(s)
+
+    # Start up the Scanners
+    for s in scanners:
+        s["proc"].start()
+
+    # Iterate through all the Scanners and see if they died
+    # Respawn them if they have.
+    # (The Ravensticks occasionally suffer glib errors)
+    try:
+        while 1:
+            for i, s in enumerate(scanners):
+                
+                # Wait on the join and then start it again if it died
+                s["proc"].join(1)
+                if not s["proc"].is_alive():
```

Comment:
Check out create_device on line 128. It fires an alarm after 10 seconds if nothing happens and then tries again. If it fails to work 5 times in a row, it throws an exception. In my experience (and from what I gathered talking to you last summer) they rarely fail more than 2-3 times in a row.